src/main/java/pl/betoncraft/betonquest/objectives/PickupObjective.java

```diff
@@ -0,0 +1,129 @@
+package pl.betoncraft.betonquest.objectives;
+
+import org.bukkit.Bukkit;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Listener;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.EventHandler;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+
+import pl.betoncraft.betonquest.BetonQuest;
+import pl.betoncraft.betonquest.Instruction;
+import pl.betoncraft.betonquest.api.Objective;
+import pl.betoncraft.betonquest.config.Config;
+import pl.betoncraft.betonquest.Instruction.Item;
+import pl.betoncraft.betonquest.utils.PlayerConverter;
+import pl.betoncraft.betonquest.exceptions.InstructionParseException;
+
+public class PickupObjective extends Objective implements Listener {
+
+    private final Item[] pickupItems;
+
+    private final int amount;
+
+    private final boolean notify;
+
+    public PickupObjective(final Instruction instruction) throws InstructionParseException {
+        super(instruction);
+
+        template = PickupData.class;
+
+        pickupItems = instruction.getItemList();
+
+        amount = instruction.getInt(instruction.getOptional("amount"), 1);
+        notify = instruction.hasArgument("notify");
+    }
+
+    @EventHandler(ignoreCancelled = true)
+    public void onPickup(final EntityPickupItemEvent event) {
+        if (!isValidItem(event.getItem().getItemStack()) || !(event.getEntity() instanceof Player)) {
+            return;
+        }
+
+        final Player player = (Player) event.getEntity();
+        final String playerID = PlayerConverter.getID(player);
+
+        if (!containsPlayer(playerID) || !checkConditions(playerID)) {
+            return;
+        }
+
+        final PickupObjective.PickupData playerData = (PickupObjective.PickupData) dataMap.get(playerID);
```

Comment:
Extract `(PickupObjective.PickupData) dataMap.get(playerID)` into its own method as it is used very often.